server/src/main/java/org/eclipse/lsp/cobol/core/visitor/CobolVisitor.java

```diff
@@ -868,7 +868,45 @@ private void areaBWarning(@NonNull List<Token> tokenList) {
   }
 
   private List<Node> addTreeNode(ParserRuleContext ctx, Function<Locality, Node> nodeConstructor) {
-    return VisitorHelper.createTreeNode(positions, visitChildren(ctx), ctx, nodeConstructor);
+    List<Node> children = visitChildren(ctx);
+    return retrieveRangeLocalityWithSameUri(ctx)
+        .map(constructNode(nodeConstructor, children))
+        .orElse(children);
+  }
+
+  private Optional<Locality> retrieveRangeLocalityWithSameUri(ParserRuleContext ctx) {
+    if (ctx == null) return Optional.empty();
```

Comment:
IMHO, the solution that looks like currying looks prettier.

Reply:
Let's see what can I offer here.